WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -393,17 +394,49 @@ public ICoinsView AsAllCoinsView()
 
 	public ICoinsView Available() => AsCoinsView().Available();
 
-	public ICoinsView ChildrenOf(SmartCoin coin) => AsCoinsView().ChildrenOf(coin);
-
 	public ICoinsView CoinJoinInProcess() => AsCoinsView().CoinJoinInProcess();
 
 	public ICoinsView Confirmed() => AsCoinsView().Confirmed();
 
-	public ICoinsView DescendantOf(SmartCoin coin) => AsCoinsView().DescendantOf(coin);
+	public ImmutableArray<SmartCoin> DescendantOf(SmartCoin coin, bool includeSelf)
+	{
+		lock (Lock)
+		{
+			return DescendantOfNoLock(coin, includeSelf);
+		}
+	}
+
+	/// <summary>Gets descendant coins of the given coin - i.e. all coins that spent the input coin, all coins that spent those coins, etc.</summary>
+	/// <remarks>Callers must acquire <see cref="Lock"/> before calling this method.</remarks>
+	private ImmutableArray<SmartCoin> DescendantOfNoLock(SmartCoin coin, bool includeSelf)
+	{
+		ICoinsView allCoins = AsAllCoinsViewNoLock();
+
+		IEnumerable<SmartCoin> Generator(SmartCoin parentCoin, bool addSelf)
+		{
+			IEnumerable<SmartCoin> childrenOf = parentCoin.SpenderTransaction is not null
+				? allCoins.Where(x => x.TransactionId == parentCoin.SpenderTransaction.GetHash()) // Inefficient.
+				: Array.Empty<SmartCoin>();
+
+			foreach (var child in childrenOf)
+			{
+				foreach (var childDescendant in Generator(child, addSelf: false))
+				{
+					yield return childDescendant;
+				}
+
+				yield return child;
+			}
 
-	private ICoinsView DescendantOfAndSelfNoLock(SmartCoin coin) => AsAllCoinsViewNoLock().DescendantOfAndSelf(coin);
+			// Return self.
+			if (addSelf)
+			{
+				yield return parentCoin;
+			}
+		}
 
-	public ICoinsView DescendantOfAndSelf(SmartCoin coin) => AsAllCoinsView().DescendantOfAndSelf(coin);
+		return Generator(coin, addSelf: includeSelf).ToImmutableArray();
```

Comment:
How come that now we use `ImmutableArray` instead of `ICoinsView` as everywhere else?

Reply:
> test that method instead.

Well, `DescendantOf` only locks then returns `DescendantOfNoLock`, so in a way it is tested. I'm fine with it, it's up to you.